man/tss2_quote.1.md
% tss2_quote(1) tpm2-tools | General Commands Manual
%
% APRIL 2019

# NAME

**tss2_quote**(1) -

# SYNOPSIS

**tss2_quote** [*OPTIONS*]

[common fapi references](common/tss2-fapi-references.md)

# DESCRIPTION

**tss2_quote**(1) - This command performs an attestation using the TPM. The PCR bank for each provided PCR index and signing scheme are set in the cryptographic profile
(cf., **fapi-profile(5)**).

# OPTIONS

These are the available options:

  * **-x**, **\--pcrList**=_STRING_:

    An array holding the PCR indices to quote against.

  * **-Q**, **\--qualifyingData**=_FILENAME_ or _-_ (for stdin):

    A nonce provided by the caller to ensure freshness of the signature. Optional parameter.

  * **-l**, **\--pcrLog**=_FILENAME_ or _-_ (for stdout):

    Returns the PCR log for the chosen PCR. Optional parameter.

    PCR event logs are a list (arbitrary length JSON array) of log entries with
    the following content.

        - recnum: Unique record number
        - pcr: PCR index
        - digest: The digests
        - type: The type of event. At the moment the only possible value is: "LINUX_IMA" (legacy IMA)
        - eventDigest: Digest of the event; e.g. the digest of the measured file
        - eventName: Name of the event; e.g. the name of the measured file.

  * **-f**, **\--force**:

    Force overwriting the output file.

  * **-p**, **\--keyPath**=_STRING_:

    Identifies the signing key.

  * **-q**, **\--quoteInfo**=_FILENAME_ or _-_ (for stdout):

    Returns a JSON-encoded structure holding the inputs to the quote operation. This includes the digest value and PCR values.

  * **-o**, **\--signature**=_FILENAME_ or _-_ (for stdout):

    Returns the signature over the quoted material.

  * **-c**, **\--certificate**=_FILENAME_ or _-_ (for stdout):

    The certificate associated with keyPath in PEM format. Optional parameter.

[common tss2 options](common/tss2-options.md)

# EXAMPLE
```
tss2_quote --keyPath=HS/SRK/quotekey --pcrList="10,16" --qualifyingData=qualifyingData.file --signature=signature.file --pcrLog=pcrLog.file --certificate=certificate.file --quoteInfo=quoteInfo.info
```

# RETURNS

0 on success or 1 on failure.

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/fapi/tss2_quote.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "tools/fapi/tss2_template.h"

/* needed by tpm2_util and tpm2_option functions */
bool output_enabled = false;

/* Context struct used to store passed command line parameters */
static struct cxt {
    uint32_t   *pcrList;
    char const *keyPath;
    char const *qualifyingData;
    char const *quoteInfo;
    char const *pcrLog;
    char const *signature;
    char const *certificate;
    bool        overwrite;
} ctx;

/**
 * Split the comma separated input, parse each token as number,
 * put the numbers in the array output.  Allocate memory for
 * output to hold the numbers.  The last element of output is -1.
 *
 * On failure returns false and output is not allocated.
 * On success the caller frees output.
 */
static inline bool extract_pcrs(char *input, uint32_t **output) {
    size_t size = 1;
    char *temp = input;
    while ((temp = strchr (temp+1, ','))) size++;
    *output = malloc (sizeof(uint32_t) * (size + 1));
    if (!*output) {
        fprintf (stderr, "malloc failed: %m\n");
        return false;
    }
    (*output)[size] = -1;
    char *x = strtok_r (input, ",", &temp);
    if (!tpm2_util_string_to_uint32(x, output[0])) {
        fprintf (stderr, "%s cannot be used as PCR\n", x);
        free (*output);
        return false;
    }
    size = 0;
    while ((x = strtok_r (NULL, ",", &temp))) {
      if (!tpm2_util_string_to_uint32(x, &(*output)[++size])) {
            fprintf (stderr, "%s cannot be used as PCR\n", x);
            free (*output);
            return false;
        }
    }
    return true;
}

/* Parse command line parameters */
static bool on_option(char key, char *value) {
    (void)value;
    switch (key) {
    case 'x':
        return extract_pcrs(value, &ctx.pcrList);
    case 'Q':
        ctx.qualifyingData = value;
        break;
    case 'l':
        ctx.pcrLog = value;
        break;
    case 'f':
        ctx.overwrite = true;
        break;
    case 'p':
        ctx.keyPath = value;
        break;
    case 'q':
        ctx.quoteInfo = value;
        break;
    case 'o':
        ctx.signature = value;
        break;
    case 'c':
        ctx.certificate = value;
        break;
    }
    return true;
}

/* Define possible command line parameters */
bool tss2_tool_onstart(tpm2_options **opts) {
    struct option topts[] = {
        {"pcrList"       , required_argument, NULL, 'x'},
        {"keyPath"        , required_argument, NULL, 'p'},
        {"qualifyingData", required_argument, NULL, 'Q'},
        {"quoteInfo"     , required_argument, NULL, 'q'},
        {"signature"      , required_argument, NULL, 'o'},
        {"pcrLog"        , required_argument, NULL, 'l'},
        {"certificate"    , required_argument, NULL, 'c'},
        {"force"          , no_argument      , NULL, 'f'}
    };
    return (*opts = tpm2_options_new ("x:Q:l:fp:q:o:c:", ARRAY_LEN(topts),
        topts, on_option, NULL, 0)) != NULL;
}

/* Execute specific tool */
int tss2_tool_onrun (FAPI_CONTEXT *fctx) {
    /* Check availability of required parameters */
    if (!ctx.pcrList) {
        fprintf (stderr, "No PCRs were chosen, use --pcrList\n");
        return -1;
    }
    if (!ctx.keyPath) {
        fprintf (stderr, "No key path provided, use --keyPath\n");
        free (ctx.pcrList);
        return -1;
    }
    if (!ctx.quoteInfo) {
        fprintf (stderr, "No quoteInfo provided, use --quoteInfo\n");
        free (ctx.pcrList);
        return -1;
    }
    if (!ctx.signature) {
        fprintf (stderr, "No signature provided, use --signature\n");
        free (ctx.pcrList);
        return -1;
    }

    /* Check exclusive access to stdout */
    int count_out = 0;
    if (ctx.quoteInfo && !strcmp (ctx.quoteInfo, "-")) count_out +=1;
    if (ctx.pcrLog && !strcmp (ctx.pcrLog, "-")) count_out +=1;
    if (ctx.signature && !strcmp (ctx.signature, "-")) count_out +=1;
    if (ctx.certificate && !strcmp (ctx.certificate, "-")) count_out +=1;
    if (count_out > 1) {
        fprintf (stderr, "Only one of --quoteInfo, --pcrLog, --signature and "\
            "--certificate can print to - (standard output)\n");
        free (ctx.pcrList);
        return -1;
    }

    /* Read qualifyingData file */
    TSS2_RC r;
    uint8_t *qualifyingData = NULL;
    size_t qualifyingDataSize = 0;
    if (ctx.qualifyingData) {
        r = open_read_and_close (ctx.qualifyingData,
            (void*)&qualifyingData, &qualifyingDataSize);
        if (r) {
          LOG_ERR ("open_read_and_close qualifyingData", r);
          free (ctx.pcrList);
          return 1;
        }
    }

    /* Execute FAPI command with passed arguments */
    uint8_t *signature;
    size_t signatureSize;
    char *quoteInfo, *pcrLog = NULL, *certificate = NULL;
    r = Fapi_Quote (fctx, ctx.pcrList, 1, ctx.keyPath,
        NULL, qualifyingData, qualifyingDataSize, &quoteInfo,
        &signature, &signatureSize, &pcrLog, &certificate);
    if (r != TSS2_RC_SUCCESS) {
        LOG_PERR ("Fapi_Quote", r);
        free (ctx.pcrList);
        free (qualifyingData);
        return 1;
    }

    free (ctx.pcrList);
    free (qualifyingData);

    /* Write returned data to file(s) */
    if (ctx.quoteInfo && quoteInfo) {
        r = open_write_and_close (ctx.quoteInfo, ctx.overwrite, quoteInfo,
            strlen(quoteInfo));
        if (r) {
            LOG_PERR ("open_write_and_close quoteInfo", r);
            Fapi_Free (quoteInfo);
            Fapi_Free (pcrLog);
            Fapi_Free (signature);
            Fapi_Free (certificate);
            return 1;
        }
    }

    Fapi_Free (quoteInfo);

    if (ctx.pcrLog && pcrLog) {
        r = open_write_and_close (ctx.pcrLog, ctx.overwrite, pcrLog,
            strlen(pcrLog));
        if (r) {
            LOG_PERR ("open_write_and_close pcrLog", r);
            Fapi_Free (pcrLog);
            Fapi_Free (signature);
            Fapi_Free (certificate);
            return 1;
        }
    }
    Fapi_Free (pcrLog);

    if (ctx.signature && signature) {
        r = open_write_and_close (ctx.signature, ctx.overwrite, signature,
            signatureSize);
        if (r) {
            LOG_PERR ("open_write_and_close signature", r);
            Fapi_Free (signature);
            Fapi_Free (certificate);
            return 1;
        }
    }
    Fapi_Free (signature);

    if (ctx.certificate && certificate) {
        r = open_write_and_close (ctx.certificate, ctx.overwrite, certificate,
            strlen(certificate));
        if (r) {
            LOG_PERR ("open_write_and_close certificate", r);
            Fapi_Free (certificate);
            return 1;
        }
    }
    Fapi_Free (certificate);

    return 0;
}
